### PR TITLE
Rework hierarchy_driver and add timer

### PR DESCRIPTION
**Pull request by Rombur**
This PR reworks hierarchy_driver so that it can be used for profiling. Also add a few timers for some crude profiling.

### CODE REVIEW

**Comment by Rombur** [comment]
_ping_

File: tests/hierarchy_driver.cc [hunk 1]
```diff
@@ -1,11 +1,88 @@
...
+  auto material_property =
+      MaterialPropertyFactory<dim>::create_material_property(
+          params->get<std::string>("material_property.type"));
+  Source<dim> source;
+
+  auto laplace_ptree = params->get_child("laplace");
+  LaplaceMatrixFree<dim, fe_degree, double> mf_laplace(comm);
+  mf_laplace.setup_system(laplace_ptree, *material_property);
+
+  auto const locally_owned_dofs = mf_laplace._locally_owned_dofs;
```
Comment (masterleinad) [inline tests/hierarchy_driver.cc]: Any reason to copy the variable?

File: tests/hierarchy_driver.cc [hunk 2]
```diff
@@ -1,11 +1,88 @@
...
+  auto timer = std::make_shared<dealii::TimerOutput>(
+      comm, pcout, dealii::TimerOutput::summary,
+      dealii::TimerOutput::wall_times);
+
+  auto material_property =
+      MaterialPropertyFactory<dim>::create_material_property(
+          params->get<std::string>("material_property.type"));
+  Source<dim> source;
+
+  auto laplace_ptree = params->get_child("laplace");
```
Comment (masterleinad) [inline tests/hierarchy_driver.cc]: Can't this be a `const &`?

File: tests/hierarchy_driver.cc [hunk 3]
```diff
@@ -1,11 +1,88 @@
...
+    hierarchy.vmult(solution, rhs);
+
+    mf_laplace._laplace_operator.vmult(residual, solution);
+    residual.sadd(-1., 1., rhs);
+    double rel_residual = residual.l2_norm() / residual0_norm;
+    res[i + 1] = rel_residual;
+  }
+
+  double const conv_rate = res[n_cycles] / res[n_cycles - 1];
+  pcout << "Convergence rate: " << std::fixed << std::setprecision(2)
```
Comment (masterleinad) [inline tests/hierarchy_driver.cc]: I guess we should aim for similar tolerances in the other tests as well...

File: tests/hierarchy_driver.cc [hunk 4]
```diff
@@ -80,38 +163,198 @@ int main(int argc, char *argv[])
...
+                    "use matrix-free algorithm");
 
   boost_po::variables_map vm;
   boost_po::store(boost_po::parse_command_line(argc, argv, cmd), vm);
   boost_po::notify(vm);
 
-  f_type filenames = {"hierarchy_input.info"};
+  if (vm.count("help"))
+  {
+    std::cout << cmd << std::endl;
```
Comment (masterleinad) [inline tests/hierarchy_driver.cc]: Do we have some kind of avoid `std::endl` policy?

File: tests/hierarchy_driver.cc [hunk 5]
```diff
@@ -80,38 +163,198 @@ int main(int argc, char *argv[])
...
     dim = vm["dim"].as<int>();
-
   mfmg::ASSERT(dim == 2 || dim == 3, "Dimension must be 2 or 3");
 
-  for (auto &filename : filenames)
+  bool matrix_free = false;
+  if (vm.count("matrix_free"))
+    matrix_free = vm["matrix_free"].as<bool>();
+
+  auto params = std::make_shared<boost::property_tree::ptree>();
```
Comment (masterleinad) [inline tests/hierarchy_driver.cc]: Can't this be `const`?

File: tests/hierarchy_driver.cc [hunk 6]
```diff
@@ -80,38 +163,198 @@ int main(int argc, char *argv[])
...
+  boost::property_tree::info_parser::read_info(filename, *params);
+
+  int fe_degree = params->get<unsigned int>("laplace.fe_degree", 1);
```
Comment (masterleinad) [inline tests/hierarchy_driver.cc]: Can't this be `const`?

File: tests/hierarchy_driver.cc (hunk 1 above)
Comment (Rombur) [reply tests/hierarchy_driver.cc]: Just to shorten the name :)

File: tests/hierarchy_driver.cc (hunk 3 above)
Comment (Rombur) [reply tests/hierarchy_driver.cc]: This is not a test. It's just an executable that we can use for profiling.

File: tests/hierarchy_driver.cc (hunk 4 above)
Comment (Rombur) [reply tests/hierarchy_driver.cc]: No, we don't

File: include/mfmg/common/hierarchy.hpp [hunk 7]
```diff
@@ -24,13 +24,28 @@
...
+#include <deal.II/base/timer.h>
+
 #include <boost/property_tree/ptree.hpp>
 
 #include <memory>
 #include <vector>
 
 namespace mfmg
 {
+void timer_enter_subsection(std::shared_ptr<dealii::TimerOutput> const &timer,
```
Comment (dalg24) [inline include/mfmg/common/hierarchy.hpp]: Why did you chose to pass a smart pointer over a reference?
When would the pointer be null?

File: tests/hierarchy_driver.cc (hunk 1 above)
Comment (dalg24) [reply tests/hierarchy_driver.cc]: Can/should it be a reference?

File: tests/hierarchy_driver.cc [hunk 8]
```diff
@@ -80,38 +163,198 @@ int main(int argc, char *argv[])
...
-  using f_type = std::vector<std::string>;
-
   MPI_Init(&argc, &argv);
 
   boost_po::options_description cmd("Available options");
-  cmd.add_options()("filename,f", boost_po::value<f_type>()->multitoken(),
-                    "file(s) containing input parameters");
+  cmd.add_options()("help,h", "produce help message");
+  cmd.add_options()("filename,f", boost_po::value<std::string>()->multitoken(),
+                    "file containing input parameters");
```
Comment (dalg24) [inline tests/hierarchy_driver.cc]: Does that let you drop the `--filename=`?

File: tests/hierarchy_driver.cc [hunk 9]
```diff
@@ -80,38 +163,198 @@ int main(int argc, char *argv[])
...
   cmd.add_options()("dim,d", boost_po::value<int>(), "dimension");
+  cmd.add_options()("matrix_free,m", boost_po::value<bool>(),
+                    "use matrix-free algorithm");
```
Comment (dalg24) [inline tests/hierarchy_driver.cc]: Why did you choose `value<bool>()` over `bool_switch()`?

File: tests/hierarchy_driver.cc [hunk 10]
```diff
@@ -80,38 +163,198 @@ int main(int argc, char *argv[])
...
+      {
+        matrix_free_two_grids<3, 10>(params);
+
+        break;
+      }
+      default:
+      {
+        mfmg::ASSERT(false, "The fe_degree should be between 1 and 10");
+      }
+      }
```
Comment (dalg24) [inline tests/hierarchy_driver.cc]: I almost want to ask for a macro...

File: include/mfmg/common/hierarchy.hpp (hunk 7 above)
Comment (Rombur) [reply include/mfmg/common/hierarchy.hpp]: The pointer will always be null except when running the `hierarchy_driver` to profile the code.

File: tests/hierarchy_driver.cc (hunk 8 above)
Comment (Rombur) [reply tests/hierarchy_driver.cc]: Not sure what you mean but `filename` is optional

File: tests/hierarchy_driver.cc (hunk 9 above)
Comment (Rombur) [reply tests/hierarchy_driver.cc]: no reason

**Comment by Rombur** [comment]
Can we merge this PR or do you want more changes?

**Comment by dalg24** [comment]
Fine by me

**Pull request merged**